File: leetcode/067-add-binary/solution.py
class Solution:
    def addBinary(self, a: str, b: str) -> str:
        res = ""
        i, j, carry = len(a) - 1, len(b) - 1, 0
        while i >= 0 or j >= 0 or carry:
            total = carry
            if i >= 0:
                total += int(a[i])
                i -= 1
            if j >= 0:
                total += int(b[j])
                j -= 1
            res = str(total % 2) + res
            carry = total // 2
        return res


if __name__ == "__main__":
    print(Solution().addBinary("11", "1"))      # 100
    print(Solution().addBinary("1010", "1011")) # 10101

File: leetcode/067-add-binary/solution.cpp
#include <bits/stdc++.h>
using namespace std;

class Solution {
public:
    string addBinary(string a, string b) {
        string res = "";
        int i = (int)a.size() - 1, j = (int)b.size() - 1, carry = 0;
        while (i >= 0 || j >= 0 || carry) {
            int sum = carry;
            if (i >= 0) sum += a[i--] - '0';
            if (j >= 0) sum += b[j--] - '0';
            res += (char)('0' + sum % 2);
            carry = sum / 2;
        }
        reverse(res.begin(), res.end());
        return res;
    }
};

int main() {
    Solution sol;
    cout << sol.addBinary("11", "1") << endl;    // 100
    cout << sol.addBinary("1010", "1011") << endl; // 10101
    return 0;
}